### ISSUE TITLE
Renaming in Jupyter notebooks only works in a single cell

### ISSUE THREAD

**Issue by ted1030** [microsoft/pylance-release#3279]
## Environment data

VS Code version: 1.71
OS version: Windows 11 22622.586 x64
Restricted Mode: No
Python version: 3.8.10
Python extension: v2022.15.12441015
Pylance extension: v2022.8.51
Jupyter extension: v2022.9.1002461039

## Code Snippet

```python
tmp = 5
tmp = 6
```
```python
tmp = 7
```
```python
def foo():
    pass
```
```python
a = foo()
```
## Repro Steps

1. Create a ipynb file and save it.
2. Try to rename the function or the variable.
3. Create a ipynb file but don't save it.
4. Try to rename the function or the variable.

## Expected behavior

In any scenario, renaming works for all cells.

## Actual behavior

In a saved ipynb file,
renaming the function "foo" works for all cells, but renaming the variable "tmp" only works for a single cell.

In an unsaved ipynb file,
renaming the function "foo" and the variable "tmp" both only works for a single cell.

## Logs

For a saved file, [tmp.ipynb.txt](https://github.com/microsoft/pylance-release/files/9481883/tmp.ipynb.txt)
For an unsaved file, [Untitled-1.ipynb.txt](https://github.com/microsoft/pylance-release/files/9481884/Untitled-1.ipynb.txt)

**Comment by debonte** [comment]
Rename in notebooks appears to effectively do the following:
1. Look from the current cell upwards until it finds the first declaration of the target symbol.
2. Renames all usages in the declaring cell and all subsequent lines (in the declaring and subsequent cells) up to, but not including, the next declaration.

So the title "Rename in Jupyter notebooks only works in a single cell" isn't quite right. Usages in other cells will be renamed, until another declaration is hit.

For example, if I initiate a rename of `old` on any of the `print(old)` lines below, the uses of `old` marked with `# Renamed` will be renamed and the others will not:

```python
# cell 0
old = 0

# cell 1
old = 1     # Renamed (in same cell as previous declaration)
old = 2     # Renamed (previous declaration)
print(old)  # Renamed

# cell 2
print(old)  # Renamed

# cell 3
print(old)  # Renamed
old = 3
```

I'm guessing this has something to do with the way that chained source files are implicitly imported.

@heejaechang, any thoughts on what the issue might be or where to look next?

**Comment by heejaechang** [comment]
I think find all reference is not considering variable defined in another virtual file as decl unlike other symbols such as func or class.

**Comment by heejaechang** [comment]
I believe these 2 are different issues.

> In a saved ipynb file,
> renaming the function "foo" works for all cells, but renaming the variable "tmp" only works for a single cell.

Here, a symbol re-defined in a cell (internally a virtual file) is hiding symbols defined in its previous cells. symbol re-defined in the same cell works since binder will hold onto all those re-definitions. (decls)

ex)
``` python
var = 1
var = 2 # redefined `var`

def foo(): pass
def foo(): pass # redefined `foo`
```

it works in the regular file case, since other file will have an explicit import statement for the imported symbol and 2 symbols in 2 different files are connected through [resolveAliasDeclaration](https://github.com/microsoft/pyrx/blob/main/packages/pyright/packages/pyright-internal/src/analyzer/aliasDeclarationUtils.ts#L26) call.

I think to fix this issue, we either need to update `resolveAliasDeclaration` to handle for chained scope or update [getDeclarationsForNode](https://github.com/microsoft/pyrx/blob/main/packages/pyright/packages/pyright-internal/src/languageService/documentSymbolCollector.ts#L100) to return chained scope's decls as well for LS features.

Difference will be `resolveAliasDeclaration` will change the core behavior so everything including type eval will be affected but `getDeclarationsForNode` will only affect LS features (at least ones that use the method to get decls for a name node)

I think we should go with `getDeclarationsForNode` for now.

> In an unsaved ipynb file,
> renaming the function "foo" and the variable "tmp" both only works for a single cell.

this is because for unsaved file, we only restrict rename scope to a current file which is 1 cell in notebook case. we should relax it for notebook I think. and allow renaming other cells even if notebook file is not saved (has untitled schema)

**Comment by heejaechang** [comment]
hmm... I just tried (untitled case - notebook file not saved), and it works as expected. so the second case might be already fixed.

**Comment by ted1030** [comment]
I've tried again in the following environment. Except for the problem mentioned by @debonte which remains in both saved and unsaved files, the other problems have been resolved.

My current environment:
VS Code 1.75
Python v2023.3.10581011
Pylance v2023.2.43
Jupyter v2023.1.2010391206


> ```python
> # cell 0
> old = 0
> 
> # cell 1
> old = 1     # Renamed (in same cell as previous declaration)
> old = 2     # Renamed (previous declaration)
> print(old)  # Renamed
> 
> # cell 2
> print(old)  # Renamed
> 
> # cell 3
> print(old)  # Renamed
> old = 3
> ```

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.3.22, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023322-15-March-2023-PreRelease)
